samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1430 @@
+ *
+ * One thing that makes these calculations complicated is that with the data provided by the mesh we move from one node to the other. But those steps (either horizontal or vertical) does not neccesarily
+ * go along the x and y axis. Because of that each vertical and horizontal step needs to be digested into x and y movement. Given that fact, for each "rectangularish" block that holds information on LOD required,
+ * there need to be 4 movements calculated and compared with their counterparts on the texture side.
+ *
+ * Naming convention explained and method:
+ * - first mentioning of either "..vertical.." or "..horizontal.." in the variable name means that this variable is used in calculations related to moving one node down (vertical) or right (horizontal) from the current position.
+ *   Calculations are handled from the top-left corner of the texture, so we are moving either to bottom or right (on the texture, not neccessarily on the screen).
+ * - pH stands for "point H". It is a separate point for the vertical and horizontal step, from which the step is splitted into x and y on-screen axis.
+ * - A is a vertice we start calculations from. From A we move to either bottom node (which is B) or to the right node (which is C).
+ *
+ * - IMPORTANT:    I assume that:
+ *						- each block is a parallelogram which is obviously not 1:1 true, but the more precise we get (the more blocks we split the texture into) the more accurate this statement is.
+ *                      - the image is not "stretched' within a single block, which has the same rules as stated above.
+ *
+ *					With those assumption, I'm providing a parralel lines from the Ph point to the corresponding edges. This creates an another parrallelogram.
+ *
+ * Variables named: ..vertical_vertical.. or ..vertical_horizontal_top.. should be understood as:
+ * This relates to the vertical step (from A --> B) and describes (the edge from pH to the corresponding vertical edge) or (describes the edge from the pH to the corresponding horizontal-top edge).
+ *
+ * Assuming that the image is not stretched within a single block, I calculate the ratio of for example (...vertical_vertical... / AB_vertical) or (...vertical_horizontal_top... / AC_horizontal).
+ * I know, that each parrallelogram on-screen corresponds to the fixed-size rectangular on-texture. Given the ratio I calculated I can just get calculate the on-texture step in texels from the right-triangle property and compare it to the x or y step of vertical/horizontal step in pixels on-screen.
+ */
+void SparseImage::CalculateMipLevelData::calculate_mip_levels()
+{
+	size_t num_rows    = mesh.size() - 1U;
+	size_t num_columns = mesh[0].size() - 1U;
+
+	mip_table.resize(num_rows);
+	for (auto &row : mip_table)
+	{
+		row.resize(num_columns);
+	}
+
+	// Single, on-texture step in texels
+	double dTu = texture_base_dim.width / (num_columns);
+	double dTv = texture_base_dim.height / (num_rows);
```

Comment:
Why do you bracket `num_columns` and `num_rows` here?

Reply:
Right. I guess it is a remaining of something like (num_columns + 1U). Will remove that.